tensorly/regression/cp_plsr.py

```diff
@@ -0,0 +1,259 @@
+        if T.ndim(Y) == 1:
+            Y = T.reshape(Y, (-1, 1))
+
+        # Mean center the data, record info the object
+        self.X_shape_ = T.shape(X)
+        self.Y_shape_ = T.shape(Y)
+        self.X_mean_ = T.mean(X, axis=0)
+        self.Y_mean_ = T.mean(Y, axis=0)
+        X -= self.X_mean_
+        Y -= self.Y_mean_
+
+        self.X_factors = [
+            T.zeros((l, self.n_components), **T.context(X)) for l in T.shape(X)
+        ]
+        self.Y_factors = [
+            T.zeros((l, self.n_components), **T.context(X)) for l in T.shape(Y)
+        ]
+
+        ## FITTING EACH COMPONENT
+        for component in range(self.n_components):
+            comp_X_factors = [ff[:, component] for ff in self.X_factors]
+            comp_Y_factors_0 = Y[:, 0]
+            old_comp_Y_factors_0 = T.ones(T.shape(comp_Y_factors_0)) * T.inf
+
+            for iter in range(self.n_iter_max):
+                Z = T.tensordot(X, comp_Y_factors_0, axes=((0,), (0,)))
+                if Z.ndim >= 2:
+                    Z_comp = tucker(Z, [1] * T.ndim(Z))[1]
+                else:
+                    Z_comp = [Z / T.norm(Z)]
+                for mode in range(1, X.ndim):  # First mode of Z is collapsed by the above tensordot call
+                    comp_X_factors[mode] = T.reshape(Z_comp[mode - 1], (-1,))
+
+                comp_X_factors[0] = multi_mode_dot(
+                    X, comp_X_factors[1:], range(1, T.ndim(X))
+                )
+                comp_Y_factors_1 = T.dot(T.transpose(Y), comp_X_factors[0])
+                comp_Y_factors_1 /= T.norm(comp_Y_factors_1)
+                comp_Y_factors_0 = T.dot(Y, comp_Y_factors_1)
+
+                if T.norm(old_comp_Y_factors_0 - comp_Y_factors_0) < self.tol:
+                    if self.verbose:
+                        print(f"Component {component}: converged after {iter} iterations")
+                    break
+                old_comp_Y_factors_0 = T.copy(comp_Y_factors_0)
+
+            # Put iteration results back to the parameter variables
+            for ii in range(len(comp_X_factors)):
+                self.X_factors[ii] = T.index_update(
+                    self.X_factors[ii], T.index[:, component], comp_X_factors[ii]
+                )
+            self.Y_factors[0] = T.index_update(
+                self.Y_factors[0], T.index[:, component], comp_Y_factors_0
+            )
+            self.Y_factors[1] = T.index_update(
+                self.Y_factors[1], T.index[:, component], comp_Y_factors_1
+            )
+
+            # Deflation
+            X -= CPTensor(
+                (None, [T.reshape(ff, (-1, 1)) for ff in comp_X_factors])
+            ).to_tensor()
+            Y -= T.dot(
+                T.dot(
+                    self.X_factors[0],
+                    T.lstsq(self.X_factors[0], T.reshape(comp_Y_factors_0, (-1, 1)))[0],
+                ),
+                T.reshape(comp_Y_factors_1, (1, -1)),
+            )  # Y -= T pinv(T) u q'
```

Comment:
The comment is really welcome here. Would it make sense to add a pinv function to TensorLy, specifying whether we want right or left?

Reply:
I could see pinv potentially being useful. Ultimately, we don't need it here. Because we're only using it once, we might as well just use lstsq.